pennylane/qnode_new/qnode.py

```diff
@@ -0,0 +1,935 @@
+        del self.queue
+        del self.obs_queue
+
+        # True if op is a CV, False if it is a discrete variable (Identity could be either)
+        are_cvs = [isinstance(op, plo.CV) for op in self.ops if not isinstance(op, plops.Identity)]
+        if not all(are_cvs) and any(are_cvs):
+            raise QuantumFunctionError(
+                "Continuous and discrete operations are not allowed in the same quantum circuit.")
+
+        # TODO: we should enforce plugins using the Device.capabilities dictionary to specify
+        # whether they are qubit or CV devices, and remove this logic here.
+        #self.type = 'cv' if all(are_cvs) else 'qubit'  #: str: circuit type, in {'cv', 'qubit'}
+
+
+    def _sort_args(self, args, kwargs):
+        """NEW: Sort the quantum function arguments to positional and keyword-only.
+
+        When the QNode is evaluated via :meth:`~QNode.__call__`, ``kwargs`` may also contain
+        differentiable POSITIONAL_OR_KEYWORD arguments given using the keyword syntax,
+        in addition to KEYWORD_ONLY arguments.
+
+        This method moves all positional arguments from ``kwargs`` to ``args``,
+        and substitutes missing positional arguments with special singleton markers.
+
+        Args:
+            args (tuple[Any]): positionally given arguments
+            kwargs (dict[str, Any]): keyword-given arguments
+
+        Returns:
+            tuple[Any], dict[str, Any]: positional arguments, auxiliary arguments
+        """
+        n_args = len(args)  # number of positionally-given args
+        pos = list(args)
+        for name, s in self.func.sig.items():
+            # is it positional?
+            if s.par.kind <= inspect.Parameter.POSITIONAL_OR_KEYWORD:
+                if s.idx < n_args:
+                    # this parameter is already given in args
+                    if name in kwargs:
+                        raise QuantumFunctionError("Quantum function parameter '{}' given twice.".format(name))
+                else:
+                    # not in args, maybe we find it in kwargs?
+                    p = kwargs.pop(name, _MARKER)  # None may be a valid parameter value, hence _MARKER
+                    pos.append(p)
+        return tuple(pos), kwargs
+
+
+    def _check_args(self, args, kwargs):
+        """NEW: Validate the quantum function arguments, apply defaults.
+
+        Here we apply all the default values for the parameters found in the signature of
+        :attr:`QNode.func`, and check if any arguments are invalid or missing.
+
+        Args:
+            args (tuple[Any]): positional (differentiable) arguments
+            kwargs (dict[str, Any]): auxiliary arguments
+
+        Returns:
+            tuple[Any], dict[str, Any]: positional arguments, auxiliary arguments
+        """
+        # pylint: disable=too-many-branches
+        # TODO maybe disallow *args in qfunc signature entirely? it complicates things and may be confusing
```

Comment:
Are `*args` currently used anywhere in our tutorials/tests?

Reply:
Not sure about tutorials, but grepping for "*arg" specifically did not result in any hits. Of course the name of the variable could be different.

Reply:
Likely culprits are `*params`, `*param`, `*pars` `*par`, and `*weights`.

Reply:
oh no :)